### PR TITLE
feature: cache credentials for direct sign-in

### LINKED ISSUE #245: Auth: direct sign-in

### PR DESCRIPTION
**Pull request by aaisel**
# Description

This PR adds the ability to store login credentials, allowing users to sign in directly without going through the `SignInScreen`. Caching is done with the help of Preferences DataStore to securely store and manage login details, enabling a seamless and efficient sign-in experience.

Fixes #245 

# How Has This Been Tested?

No tests added.
  
# Dependencies

Added necessary dependencies.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "auth")
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt:12]: Mixing it up with datastore! Good implementation!

File: app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  companion object {
+    private val USER_EMAIL = stringPreferencesKey("user_email")
+  }
+
+  // Get User Email
+  val userEmailFlow: Flow<String?> =
+      context.dataStore.data.map { preferences -> preferences[USER_EMAIL] }
+
+  /** Save the user's email in the data store */
+  suspend fun saveCredentials(userEmail: String) {
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt:24]: Document the parameter!

File: app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+
+val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "auth")
+
+class AuthManager(private val context: Context) {
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/cache/AuthenticatorDataStore.kt:14]: Document the class

File: app/src/main/java/com/github/onlynotesswent/model/authentication/Authenticator.kt [hunk 4]
```diff
@@ -83,6 +86,7 @@ open class Authenticator(
   suspend fun signOut() {
     Firebase.auth.signOut()
     credentialManager.clearCredentialState(ClearCredentialStateRequest())
+    authManager.deleteCredentials()
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/model/authentication/Authenticator.kt:89]: Nice catch of signing it out!

File: app/src/main/java/com/github/onlynotesswent/MainActivity.kt [hunk 5]
```diff
@@ -89,132 +93,161 @@ fun OnlyNotesApp(
...
 
-  NavHost(navController = navController, startDestination = Route.AUTH) {
-    navigation(
-        startDestination = Screen.AUTH,
-        route = Route.AUTH,
-    ) {
-      composable(Screen.AUTH) { SignInScreen(navigationActions, userViewModel, authenticator) }
-      composable(Screen.CREATE_USER) { CreateUserScreen(navigationActions, userViewModel) }
-    }
+  val isInitialized = remember { mutableStateOf(false) }
```
Comment (santiagoarayaramos) [inline app/src/main/java/com/github/onlynotesswent/MainActivity.kt:96]: Given this is a remember, guarantee that it is reset after we log out (and thus go back to auth screen)

**Review by santiagoarayaramos — CHANGES_REQUESTED** [review]
Nice PR! Code was clean, nice experimentation with datastore other than room, it seemed simple. This goes against consistency but we are still students trying to learn, and if it turns out well we can potentially replace room.

Some small changes request but good pr!

File: app/src/main/java/com/github/onlynotesswent/MainActivity.kt (hunk 5 above)
Comment (aaisel) [reply app/src/main/java/com/github/onlynotesswent/MainActivity.kt:96]: `isInitialized` is used only in the beginning when the app is first launched, and its purpose is to delay the navigation until the `LaunchedEffect` has finished. Afterwards, it is not needed anymore

**Review by santiagoarayaramos — APPROVED** [review]
Very nice implementation of changes! Testing it also was great, a very good addition to the app, good work!

**Pull request merged**